### ISSUE TITLE
No way to add TS helper scripts to `<head>`

### ISSUE THREAD

**Issue by AaronBeaudoin** [brillout/vite-plugin-ssr#464]
### Description

I have a TypeScript browser-only helper scripts I'd like to add in the `<head>` on my pages.

In `_default.page.server.ts` I have this:

```js
import hoverHelper from "/assets/helpers/hover?url";
import viewportHelper from "/assets/helpers/viewport?url";

export async function render(pageContext: PageContext) {
  ...

  return escape`
    <!DOCTYPE html>
    <html lang="en">
      <head>

        ...

        <script src="${hoverHelper}"></script>
        <script src="${viewportHelper}"></script>

      </head>
      <body>

        ...

      </body>
    </html>
  `;
}
```

In production however, I get an error because my page can't find either helper. It look like the build output is like this:

```js
// _default.page.server.5419f7e5.js
const hoverHelper = "/assets/hover.1b1ef52d.ts";
const viewportHelper = "/assets/viewport.2a0ad6e3.js";
```

But those two assets don't actually exist. The files don't exist:
- `dist/client/assets/hover.1b1ef52d.ts`
- `dist/client/assets/viewport.2a0ad6e3.js`

Am I not doing this correctly? How would I go about adding a TypeScript helper script to the `<head>` of my page? If it was just JS I could put it at `public/` and then just hardcode the URL and be done with it, but since I want better type safety I have to deal with the fact that my script need to go through the build process. Is this possible?

**Comment by AaronBeaudoin** [comment]
The reason I want the scripts in the head is so that they run ASAP to prevent flashing or layout shift. One of them is responsible for toggling hover effects based on whether the user is using mouse or touch input, and the other one is responsible for creating some CSS custom properties for sizing large page sections, which would cause those areas to awkwardly shift later if not done before parsing the `<body>`.

**Comment by brillout** [comment]
See https://vite-plugin-ssr.com/includeAssetsImportedByServer.

**Comment by AaronBeaudoin** [comment]
The linked feature doesn't help with this use case at all. In my case, I'm importing the asset URL and manually adding it to my HTML. I'm not actually importing the asset on the server, I'm just getting the URL.

So in my case, if I set `includeAssetsImportedByServer: true`, the files I need still are not built.

**Comment by brillout** [comment]
If you do the same but from a `.page.js` or `.page.client.js` file, does it work then?

**Comment by AaronBeaudoin** [comment]
The nature of what I'm trying to do doesn't make sense in a `.page.js` or `.page.client.js` file, so unfortunately no.

I'm trying to **load a script in the `<head>` of my document**, so I have to do that in `_default.page.server.js`. I'm not even trying to load the asset on the server, I'm just wanting to give the HTML template the right URL so the client can load it later.

**Comment by AaronBeaudoin** [comment]
My workaround for now is to just not us TS and instead put my script as a JS file in `public/` which I can then easily hardcode in my HTML string template like `<script src="/script-name.js"></script>`

**Comment by brillout** [comment]
If you `import "/assets/helpers/hover?url"` and/or `import "/assets/helpers/hover"` in `_default.page.client.js` **in addtion** to using them in `_default.page.server.js` (without actually using these in `_default.page.client.js`), does it work then? I'm asking because Vite needs a reference to all assets on the client-side.

**Comment by AaronBeaudoin** [comment]
When I add `import "/assets/helpers/hover"` in `_default.page.client.js` it includes the code in `dist/client/assets/pages/_default/_default.page.client.<hash>.js`, but that doesn't help me because the URL in my `<head>` is still something like `/assets/hover.<hash>.ts`.

If I do `import "/assets/helpers/hover?url"` in `_default.page.client.js` then I actually get a `/assets/hover.<hash>.ts` file outputted in my `dist/client/`. Problem is, it's still a TS file, so my browser isn't going to be able to run it.

So the main problem of this issue still stands. It doesn't appear that I can compile a TS file to a JS file and use it in the `<head>` of my HTML without some separate custom build step. Is that by design?

**Comment by brillout** [comment]
Yes that may currently not be possible with Vite. It's a Vite thing; create a feature request over there.

I'm leaving this ticket open because `includeAssetsImportedByServer` should make the `.ts` asset be included.

**Comment by brillout** [comment]
The vps side of this is fixed.

I tried and I can confirm that Vite doesn't seem to support `?url` importing `.ts` files.

**Comment by brillout** [comment]
Fix released in `0.4.40`.
